ecs-cli/modules/clients/aws/servicediscovery/client.go

```diff
@@ -0,0 +1,139 @@
+// permissions and limitations under the License.
+
+// Package servicediscovery contains functions for working with the route53 APIs
+// that back ECS Service Discovery
+package servicediscovery
+
+import (
+	"github.com/aws/amazon-ecs-cli/ecs-cli/modules/clients"
+	"github.com/aws/amazon-ecs-cli/ecs-cli/modules/config"
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/route53"
+	"github.com/aws/aws-sdk-go/service/servicediscovery"
+)
+
+// FindPrivateNamespace returns the ID(s) of the private namespace with the given name and vpc
+func FindPrivateNamespace(name string, vpc string, params *config.CommandConfig) ([]*string, error) {
+	region := aws.StringValue(params.Session.Config.Region)
+	var namespacesWithName []*string
+	err := listNamespaces(false, params, func(namespace *servicediscovery.NamespaceSummary) bool {
+		if aws.StringValue(namespace.Name) == name {
+			namespacesWithName = append(namespacesWithName, namespace.Id)
+		}
+		return true
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	var matchingNamespaces []*string
+	for _, namespaceID := range namespacesWithName {
+		hasVPC, err := checkNamespaceVPC(namespaceID, vpc, region, params)
+		if err != nil {
+			return nil, err
+		}
+		if hasVPC {
+			matchingNamespaces = append(matchingNamespaces, namespaceID)
+		}
+	}
+
+	return matchingNamespaces, nil
+}
+
+func checkNamespaceVPC(namespaceID *string, vpc string, region string, params *config.CommandConfig) (bool, error) {
+	namespaceInfo, err := getNamespace(namespaceID, params)
+	if err != nil {
+		return false, err
+	}
+	hostedZoneID := namespaceInfo.Namespace.Properties.DnsProperties.HostedZoneId
+	hostedZone, err := getHostedZone(hostedZoneID, params)
+	if err != nil {
+		return false, err
+	}
+	for _, hostedZoneVPC := range hostedZone.VPCs {
+		// The VPC must be in the region that we will be launching the ECS Service
+		if (aws.StringValue(hostedZoneVPC.VPCId) == vpc) && (aws.StringValue(hostedZoneVPC.VPCRegion) == region) {
+			return true, nil
+		}
+	}
+
+	return false, nil
+}
+
+// ServiceDiscovery API Actions that can be mocked in unit tests
+var sdClient *servicediscovery.ServiceDiscovery
+
+func getSDClient(params *config.CommandConfig) *servicediscovery.ServiceDiscovery {
+	if sdClient == nil {
+		sdClient = servicediscovery.New(params.Session)
+		sdClient.Handlers.Build.PushBackNamed(clients.CustomUserAgentHandler())
+	}
+	return sdClient
+}
+
+// Lists namespaces, calling 'fn' on each namespace returned.
+// To stop iterating over namespaces, return false from 'fn'
+var listNamespaces = func(isPublic bool, params *config.CommandConfig, fn func(*servicediscovery.NamespaceSummary) bool) error {
+	client := getSDClient(params)
+
+	typeFilter := servicediscovery.NamespaceTypeDnsPrivate
+	if isPublic {
+		typeFilter = servicediscovery.NamespaceTypeDnsPublic
+	}
+	request := &servicediscovery.ListNamespacesInput{
+		Filters: []*servicediscovery.NamespaceFilter{
+			&servicediscovery.NamespaceFilter{
```

Comment:
If the service discovery API already supports Filtering, is it possible to use that instead of passing in the anonymous function argument that seems to be filtering on the list of namespaces?

Reply:
Unfortunately it isn't, the API only supports filtering on type- Public or Private

Reply:
Hmm, it still seems strange to pass in an anonymous function this way if listNamespaces is meant to be a thin wrapper for the actual service discovery API. I also see we filter again on these filtered results, and it's unclear to me why we can't do it in one pass. E.g have on call to just return the results from the SDS, and then apply filtering on that as necessary after once.

Reply:
Are you suggesting something like:

```
allNamespaces := listAllNamespaces()

for _, namespace := range allNamespaces {
  // filter by name, vpc, and region
}
```

That would work; but the way I have the code right now also works, and I don't see a need to spend the time to change it to the above. 

Also, in the `FindPublicNamespace`, I use the anonymous function to terminate the iteration- this potentially saves API calls, because we can stop iterating once we have found the namespace that we want.

Also, one key thing to note, `listNamespaces` is a thin wrapper over the SDK's [ListNamespacePages](https://docs.aws.amazon.com/sdk-for-go/api/service/servicediscovery/#ServiceDiscovery.ListNamespaces) which takes an anonymous function like mine does.

Reply:
Ah, I see how the anonymous function works now -- thanks for that link. It wasn't immediately clear from the way it was written that that was what was happening.